File: telegram-bot-core/src/main/java/ru/tggc/telegrambotcore/access/annotationprovider/AnnotationProviderFactory.java
package ru.tggc.telegrambotcore.access.annotationprovider;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import ru.tggc.telegrambotcore.annotation.handle.HandleMeta;

import java.lang.reflect.Method;
import java.util.List;

@Component
@RequiredArgsConstructor
public class AnnotationProviderFactory {
    private final List<AnnotationProvider> annotationProviders;

    public HandleMeta getAnnotationMeta(Method method) {
        return annotationProviders.stream()
                .filter(ap -> ap.supports(method))
                .findFirst()
                .map(ap -> ap.extractMeta(method))
                .orElseThrow(() -> new RuntimeException("No annotation provider found for method " + method.getName()));
    }
}

File: telegram-bot-core/src/main/kotlin/ru/tggc/telegrambotcore/access/annotationprovider/AnnotationProviderFactory.kt
package ru.tggc.telegrambotcore.access.annotationprovider

import org.springframework.stereotype.Component
import ru.tggc.telegrambotcore.annotation.handle.HandleMeta
import java.lang.reflect.Method

@Component
class AnnotationProviderFactory(private val annotationProviders: MutableList<AnnotationProvider>) {
    fun getAnnotationMeta(method: Method): HandleMeta =
        annotationProviders.firstOrNull { it.supports(method) }
            ?.let { ap: AnnotationProvider -> ap.extractMeta(method) }
            ?: throw RuntimeException("AnnotationProvider not found for method: $method")
}
